packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -0,0 +1,212 @@
+            realVsizeCredentials,
+            zeroAmountCredentials,
+            zeroVsizeCredentials,
+            { signal, baseUrl: coordinatorUrl, identity: Date.now().toString(), delay: 0 },
+        );
+
+        amountCredentials = await middleware.getCredentials(
+            round.amountCredentialIssuerParameters,
+            joinedIssuance.realAmountCredentials,
+            realAmountCredentials.credentialsResponseValidation,
+            { signal, baseUrl: middlewareUrl },
+        );
+        vsizeCredentials = await middleware.getCredentials(
+            round.vsizeCredentialIssuerParameters,
+            joinedIssuance.realVsizeCredentials,
+            realVsizeCredentials.credentialsResponseValidation,
+            { signal, baseUrl: middlewareUrl },
+        );
+        outputSize = Math.max(outputSize, current.outputSize);
+    }
+
+    return {
+        accountKey,
+        amountCredentials,
+        vsizeCredentials,
+        outputSize,
+    };
+};
+
+export const getOutputAmounts = async (
+    round: CoinjoinRound,
+    accountKey: string,
+    availableVsize: number,
+    outputSize: number,
+    options: CoinjoinRoundOptions,
+) => {
+    const { roundParameters } = round;
+    const { signal, middlewareUrl } = options;
+
+    const registeredInputs = getRoundEvents('InputAdded', round.coinjoinState.events);
+    const internalAmounts: number[] = [];
+    const externalAmounts: number[] = [];
+
+    registeredInputs.forEach(i => {
+        const internal = round.inputs.find(
+            input =>
+                input.accountKey === accountKey && compareOutpoint(input.outpoint, i.coin.outpoint),
+        );
+        if (internal) {
+            internalAmounts.push(internal.confirmedAmountCredentials![0].value);
+        } else {
+            const size = getExternalOutputSize(i.coin.txOut.scriptPubKey);
+            const miningFee = Math.floor((size * roundParameters.miningFeeRate) / 1000);
+            const coordinatorFee = Math.floor(
+                roundParameters.coordinationFeeRate.rate * i.coin.txOut.value,
+            );
+            externalAmounts.push(i.coin.txOut.value - coordinatorFee - miningFee);
+        }
+    });
+    const outputAmounts = await middleware.decomposeAmounts(
+        {
+            feeRate: roundParameters.miningFeeRate,
+            allowedOutputAmounts: roundParameters.allowedOutputAmounts,
+        },
+        outputSize,
+        availableVsize, // TODO: sum all available vsize and use chunks in output registration (increase nr of outputs)
+        internalAmounts,
+        externalAmounts,
+        { signal, baseUrl: middlewareUrl },
+    );
+    options.log(`Decompose amounts: ${outputAmounts.join('')}`);
+    return outputAmounts.map(amount => {
+        const miningFee = Math.floor((outputSize * roundParameters.miningFeeRate) / 1000);
+        const coordinatorFee = 0; // TODO: should be Math.floor(roundParameters.coordinationFeeRate.rate * amount);
```

Comment:
why isn't this TODO resolved right away?

Reply:
because it's a coordinator/middleware bug we've found
it should be done like that but it isnt (will not work) i has come comment regarding that, maybe was lost during rebases.
i'll try to find it